plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3ClientSettings.java

```diff
@@ -0,0 +1,183 @@
+    final Protocol protocol;
+
+    /** An optional proxy host that requests to s3 should be made through. */
+    final String proxyHost;
+
+    /** The port number the proxy host should be connected on. */
+    final int proxyPort;
+
+    // these should be "secure" yet the api for the s3 client only takes String, so storing them
+    // as SecureString here won't really help with anything
+    /** An optional username for the proxy host, for basic authentication. */
+    final String proxyUsername;
+
+    /** An optional password for the proxy host, for basic authentication. */
+    final String proxyPassword;
+
+    /** The read timeout for the s3 client. */
+    final int readTimeoutMillis;
+
+    private S3ClientSettings(BasicAWSCredentials credentials, String endpoint, Protocol protocol,
+                             String proxyHost, int proxyPort, String proxyUsername,
+                             String proxyPassword, int readTimeoutMillis) {
+        this.credentials = credentials;
+        this.endpoint = endpoint;
+        this.protocol = protocol;
+        this.proxyHost = proxyHost;
+        this.proxyPort = proxyPort;
+        this.proxyUsername = proxyUsername;
+        this.proxyPassword = proxyPassword;
+        this.readTimeoutMillis = readTimeoutMillis;
+    }
+
+    /**
+     * Load all client settings from the given settings.
+     *
+     * Note this will always at least return a client named "default".
+     */
+    static Map<String, S3ClientSettings> load(Settings settings) {
+        Set<String> clientNames = settings.getGroups(PREFIX).keySet();
+        Map<String, S3ClientSettings> clients = new HashMap<>();
+        for (String clientName : clientNames) {
+            clients.put(clientName, getClientSettings(settings, clientName));
+        }
+        if (clients.containsKey("default") == false) {
+            // this won't find any settings under the default client,
+            // but it will pull all the fallback static settings
+            clients.put("default", getClientSettings(settings, "default"));
+        }
+        return Collections.unmodifiableMap(clients);
+    }
+
+    // pkg private for tests
+    /** Parse settings for a single client. */
+    static S3ClientSettings getClientSettings(Settings settings, String clientName) {
+        try (SecureString accessKey = getConfigValue(settings, clientName, ACCESS_KEY_SETTING, S3Repository.Repositories.KEY_SETTING);
+             SecureString secretKey = getConfigValue(settings, clientName, SECRET_KEY_SETTING, S3Repository.Repositories.SECRET_SETTING);
+             SecureString proxyUsername = getConfigValue(settings, clientName, PROXY_USERNAME_SETTING, CLOUD_S3.PROXY_USERNAME_SETTING);
+             SecureString proxyPassword = getConfigValue(settings, clientName, PROXY_PASSWORD_SETTING, CLOUD_S3.PROXY_PASSWORD_SETTING)) {
+            BasicAWSCredentials credentials = null;
+            if (accessKey.length() != 0) {
```

Comment:
I think this block would be simpler as:

```java
if (accessKey.length() == 0) {
    throw new IllegalArgumentException("Missing access key for s3 client [" + clientName + "]");
}
if (secretKey.length() == 0) {
    throw new IllegalArgumentException("Missing secret key for s3 client [" + clientName + "]");
}
final BasicAWSCredentials credentials = new BasicAWSCredentials(accessKey.toString(), secretKey.toString());
return new S3ClientSettings(...);
```

Instead of the nested if statements and double-negative-ish `if` statements